packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -0,0 +1,400 @@
+  final String textStart = originalText.substring(0, start);
+  final String textEnd = originalText.substring(end, originalText.length);
+  final String newText = textStart + replacementText + textEnd;
+  return newText;
+}
+
+/// A way to disambiguate a [TextEditingDelta] when a delta generated for an insertion
+/// and a deletion both contain collapsed [TextEditingDelta.deltaRange]'s.
+enum TextEditingDeltaType {
+  /// {@template flutter.services.TextEditingDeltaInsertion}
+  /// The delta is inserting a single/or contigous sequence of characters.
+  /// {@endtemplate}
+  insertion,
+
+  /// {@template flutter.services.TextEditingDeltaDeletion}
+  /// The delta is deleting a single/or contiguous sequence of characters.
+  /// {@endtemplate}
+  deletion,
+
+  /// {@template flutter.services.TextEditingDeltaReplacement}
+  /// The delta is replacing a range of characters with a new sequence of text.
+  ///
+  /// The range that is being replaced can either grow or shrink based on the
+  /// given replacement text.
+  ///
+  /// A replacement can occur in cases such as auto-correct, suggestions, and
+  /// when a selection is replaced by a single character.
+  /// {@endtemplate}
+  replacement,
+
+  /// {@template flutter.services.TextEditingDeltaNonTextUpdate}
+  /// The delta is not modifying the text. There are potentially selection and
+  /// composing region updates in the delta that still need to be applied to your
+  /// text model.
+  ///
+  /// A situation where this delta would be created is when dragging the selection
+  /// handles. There are no changes to the text, but there are updates to the selection
+  /// and potentially the composing region as well.
+  /// {@endtemplate}
+  nonTextUpdate,
+}
+
+/// A structure representing a granular change that has occurred to the editing
+/// state as a result of text editing.
+abstract class TextEditingDelta {
+  /// Creates a delta for a given change to the editing state.
+  ///
+  /// {@template flutter.services.TextEditingDelta}
+  /// The [oldText], [deltaText], [deltaRange], [selection], and [composing]
+  /// arguments must not be null.
+  /// {@endtemplate}
+  const TextEditingDelta({
+    required this.oldText,
+    this.deltaText = '',
+    this.deltaRange = TextRange.empty,
+    required this.selection,
+    required this.composing,
+  }) : assert(oldText != null),
+       assert(deltaText != null),
+       assert(deltaRange != null),
+       assert(selection != null),
+       assert(composing != null);
+
+  /// Creates an instance of this class from a JSON object by inferring the
+  /// type of delta based on values sent from the engine.
+  factory TextEditingDelta.fromJSON(Map<String, dynamic> encoded) {
+    final String oldText = encoded['oldText'] as String;
+    final int start = encoded['deltaStart'] as int;
+    final int end = encoded['deltaEnd'] as int;
+    final String tb = encoded['deltaText'] as String;
```

Comment:
Not a fan of the `tb` abbreviation. As far as I can tell, this naming is coming from the android embedding `replace()` parameters? 

Prefer to just fully spell out names.

Reply:
updated
``` 
start -> replacementDestinationStart
end -> replacementDestinatioEnd
tb -> replacementSource
tbStart -> replacementSourceStart
tbEnd -> replacementSourceEnd
```